File: src/Gweb.Shared/Errors/HttpErrorMapper.cs
using Gweb.Shared.Correlation;
using Gweb.Shared.Logging;
using Microsoft.AspNetCore.Http;

namespace Gweb.Shared.Errors;

public sealed record ApiErrorDetail(string Code, string Message, object? Details);

public sealed record ApiErrorBody(ApiErrorDetail Error, string? CorrelationId);

/// <summary>
/// The one place domain errors become HTTP responses. Endpoints never build error
/// bodies themselves. Never leaks stack traces, table/bucket names, or ARNs to the
/// client — those go to CloudWatch only, keyed by correlation ID.
/// </summary>
public static class HttpErrorMapper
{
    public static IResult Map(Exception error, StructuredLogger logger)
    {
        var correlationId = CorrelationScope.GetCurrent()?.CorrelationId;

        if (error is DomainException domainException)
        {
            var body = new ApiErrorBody(
                new ApiErrorDetail(domainException.Code, domainException.Message, domainException.Details),
                correlationId);
            return Results.Json(body, statusCode: domainException.HttpStatus);
        }

        logger.Error("unhandled_error", new { errorName = error.GetType().Name, errorMessage = error.Message });

        var fallbackBody = new ApiErrorBody(
            new ApiErrorDetail("INTERNAL_ERROR", "An unexpected error occurred.", null),
            correlationId);
        return Results.Json(fallbackBody, statusCode: 500);
    }
}

File: shared/errors/httpErrorMapper.ts
import { getCorrelationContext } from "../correlation/correlationContext";
import { DomainError } from "./domainErrors";
import { logger } from "../logger/logger";

export interface ApiErrorBody {
  error: { code: string; message: string; details?: unknown };
  correlationId: string | undefined;
}

export interface ApiResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

/**
 * The one place domain errors become HTTP responses. Handlers never build error
 * bodies themselves. Never leaks stack traces, table/bucket names, or ARNs to the
 * client — those go to CloudWatch only, keyed by correlation ID.
 */
export function mapErrorToHttpResponse(error: unknown): ApiResponse {
  const correlationId = getCorrelationContext()?.correlationId;

  if (error instanceof DomainError) {
    const body: ApiErrorBody = {
      error: { code: error.code, message: error.message, details: error.details },
      correlationId,
    };
    return jsonResponse(error.httpStatus, body);
  }

  logger.error({ event: "unhandled_error", ...serializeUnknownError(error) });

  const body: ApiErrorBody = {
    error: { code: "INTERNAL_ERROR", message: "An unexpected error occurred." },
    correlationId,
  };
  return jsonResponse(500, body);
}

function jsonResponse(statusCode: number, body: unknown): ApiResponse {
  return {
    statusCode,
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  };
}

function serializeUnknownError(error: unknown): { errorName?: string; errorMessage?: string } {
  if (error instanceof Error) {
    return { errorName: error.name, errorMessage: error.message };
  }
  return { errorMessage: String(error) };
}
